diff --git a/misc/logscanner2.py b/misc/logscanner2.py
index 62ddc63..18d06ac 100755
--- a/misc/logscanner2.py
+++ b/misc/logscanner2.py
@@ -107,7 +107,7 @@ def doline(line):
     if len(urlparts.query) == 0:
         return None
 
-    qsl = [ x.split('=', 1) for x in urlparts.query.split('&') ]
+    qsl = parse_qsl(urlparts.query, keep_blank_values=True)
 
     target = None
     for k,v in qsl:
