diff --git a/misc/logscanner2.py b/misc/logscanner2.py
index 62ddc63..c4ec7d8 100755
--- a/misc/logscanner2.py
+++ b/misc/logscanner2.py
@@ -53,6 +53,10 @@ def parse_apache(line):
         (method, uri, protocol) = mo.group(5).split(' ', 2)
     except ValueError:
         (method, uri, protocol) = ('-', '-', '-')
+    try:
+        body_bytes_sent = int(mo.group(7))
+    except ValueError:
+        return None
     data = {
         'remote_addr': mo.group(1),
         'time_iso8601': time_iso,
@@ -64,7 +68,7 @@ def parse_apache(line):
         'request_time': '',
         'status': mo.group(6),
         'bytes_sent': '',
-        'body_bytes-sent': int(mo.group(7)),
+        'body_bytes-sent': body_bytes_sent,
         'http_referrer': mo.group(8),
         'http_user_agent': mo.group(9),
         'ssl_cipher': '',
