diff --git a/misc/logscanner2.py b/misc/logscanner2.py
index 62ddc63..b30ab65 100755
--- a/misc/logscanner2.py
+++ b/misc/logscanner2.py
@@ -48,7 +48,10 @@ def parse_apache(line):
     mo = apachelogre.match(line)
     if not mo:
         return None
-    (time_iso, timestamp) = parse_date(mo.group(4))
+    try:
+        (time_iso, timestamp) = parse_date(mo.group(4))
+    except (KeyError, ValueError, IndexError):
+        return None
     try:
         (method, uri, protocol) = mo.group(5).split(' ', 2)
     except ValueError:
