diff --git a/src/libinjection_xss.c b/src/libinjection_xss.c
index ee12e5f..a55bce1 100644
--- a/src/libinjection_xss.c
+++ b/src/libinjection_xss.c
@@ -1000,12 +1000,9 @@ static attribute_t is_black_attr(const char *s, size_t len) {
             size_t s_without_on_len = len - 2; // temporary length variable
             while (black->name != NULL) {
                 size_t black_name_len = strlen(black->name);
-                // determine the maximum length to compare
-                size_t max_len = (s_without_on_len < black_name_len)
-                                     ? s_without_on_len
-                                     : black_name_len;
-                if (cstrcasecmp_with_null(black->name, s_without_on, max_len) ==
-                    0) {
+                if (s_without_on_len == black_name_len &&
+                    cstrcasecmp_with_null(black->name, s_without_on,
+                                          black_name_len) == 0) {
                     /* printf("Got banned attribute name %s\n", black->name); */
                     return black->atype;
                 }
