diff --git a/misc/nullserver.py b/misc/nullserver.py
index ecdc27b..d56e67d 100755
--- a/misc/nullserver.py
+++ b/misc/nullserver.py
@@ -13,10 +13,13 @@ import tornado.ioloop
 import tornado.web
 import libinjection
 
+fd = None
+
 class ShutdownHandler(tornado.web.RequestHandler):
     def get(self):
         global fd
-        fd.close()
+        if fd is not None:
+            fd.close()
         sys.exit(0)
