diff --git a/src/html5_cli.c b/src/html5_cli.c
index 275e29c..ca929be 100644
--- a/src/html5_cli.c
+++ b/src/html5_cli.c
@@ -169,6 +169,10 @@ int main(int argc, const char *argv[]) {
 
         } else if (strcmp(argv[offset], "-f") == 0) {
             offset += 1;
+            if (offset >= argc) {
+                fprintf(stderr, "usage: %s [-u] [-f flags] input\n", argv[0]);
+                return 1;
+            }
             flag = atoi(argv[offset]);
             offset += 1;
         } else {
