diff --git a/lib/std/Io/Dispatch.zig b/lib/std/Io/Dispatch.zig
index 259a233d57..5d44c27f48 100644
--- a/lib/std/Io/Dispatch.zig
+++ b/lib/std/Io/Dispatch.zig
@@ -1982,6 +1982,7 @@ fn batchCancel(userdata: ?*anyopaque, batch: *Io.Batch) void {
         const pending = &storage.pending;
         const operation_userdata: *BatchOperationUserdata = .fromErased(&pending.userdata);
         assert(operation_userdata.batch == batch);
+        index = pending.node.next;
         operation_userdata.source.cancel();
     }
     const queue: c.dispatch.queue_t = @ptrCast(batch.userdata orelse return);
