diff --git a/lib/std/Io/Dispatch.zig b/lib/std/Io/Dispatch.zig
index 259a233d57..523291f8f5 100644
--- a/lib/std/Io/Dispatch.zig
+++ b/lib/std/Io/Dispatch.zig
@@ -4276,6 +4276,7 @@ fn spawn(ev: *Evented, options: process.SpawnOptions) process.SpawnError!Spawned
 
     const pid: c.pid_t = @intCast(pid_result); // We are the parent.
     errdefer comptime unreachable; // The child is forked; we must not error from now on
+    errdefer if (prog_pipe[0] != -1) closeFd(prog_pipe[0]);
 
     fileClose(ev, err_pipe[1..2]); // make sure only the child holds the write end open
