diff --git a/lib/std/Build.zig b/lib/std/Build.zig
index f0eae560d4..3f844d98c5 100644
--- a/lib/std/Build.zig
+++ b/lib/std/Build.zig
@@ -2186,7 +2186,7 @@ fn userLazyPathsAreTheSame(lhs_lp: LazyPath, rhs_lp: LazyPath) bool {
             const rhs_sp = rhs_lp.src_path;
 
             if (lhs_sp.owner != rhs_sp.owner) return false;
-            if (std.mem.eql(u8, lhs_sp.sub_path, rhs_sp.sub_path)) return false;
+            if (!std.mem.eql(u8, lhs_sp.sub_path, rhs_sp.sub_path)) return false;
         },
         .generated => |lhs_gen| {
             const rhs_gen = rhs_lp.generated;
