diff --git a/lib/std/Build.zig b/lib/std/Build.zig
index f0eae560d4..ade04b20c8 100644
--- a/lib/std/Build.zig
+++ b/lib/std/Build.zig
@@ -2193,7 +2193,7 @@ fn userLazyPathsAreTheSame(lhs_lp: LazyPath, rhs_lp: LazyPath) bool {
 
             if (lhs_gen.file != rhs_gen.file) return false;
             if (lhs_gen.up != rhs_gen.up) return false;
-            if (std.mem.eql(u8, lhs_gen.sub_path, rhs_gen.sub_path)) return false;
+            if (!std.mem.eql(u8, lhs_gen.sub_path, rhs_gen.sub_path)) return false;
         },
         .cwd_relative => |lhs_rel_path| {
             const rhs_rel_path = rhs_lp.cwd_relative;
