diff --git a/lib/std/Build/Fuzz.zig b/lib/std/Build/Fuzz.zig
index 9d3551a565..30e54bd3da 100644
--- a/lib/std/Build/Fuzz.zig
+++ b/lib/std/Build/Fuzz.zig
@@ -69,6 +69,8 @@ const CoverageMap = struct {
     fn deinit(cm: *CoverageMap, gpa: Allocator) void {
         std.posix.munmap(cm.mapped_memory);
         cm.coverage.deinit(gpa);
+        gpa.free(cm.source_locations);
+        cm.entry_points.deinit(gpa);
         cm.* = undefined;
     }
 };
@@ -146,6 +148,10 @@ pub fn deinit(fuzz: *Fuzz) void {
     const io = fuzz.io;
     fuzz.group.cancel(io);
     fuzz.prog_node.end();
+    for (fuzz.coverage_files.values()) |*coverage_map| {
+        coverage_map.deinit(fuzz.gpa);
+    }
+    fuzz.coverage_files.deinit(fuzz.gpa);
     fuzz.gpa.free(fuzz.run_steps);
 }
