diff --git a/lib/std/Build/Fuzz.zig b/lib/std/Build/Fuzz.zig
index 9d3551a565..28169a410a 100644
--- a/lib/std/Build/Fuzz.zig
+++ b/lib/std/Build/Fuzz.zig
@@ -269,11 +269,10 @@ pub fn sendUpdate(
     // TODO: handle multiple fuzz steps in the WebSocket packets
     const coverage_map = &coverage_maps[0];
     const cov_header: *const abi.SeenPcsHeader = @ptrCast(coverage_map.mapped_memory[0..@sizeOf(abi.SeenPcsHeader)]);
-    // TODO: this isn't sound! We need to do volatile reads of these bits rather than handing the
-    // buffer off to the kernel, because we might race with the fuzzer process[es]. This brings the
-    // whole mmap strategy into question. Incidentally, I wonder if post-writergate we could pass
-    // this data straight to the socket with sendfile...
     const seen_pcs = cov_header.seenBits();
+    const seen_pcs_copy = try fuzz.gpa.alloc(usize, seen_pcs.len);
+    defer fuzz.gpa.free(seen_pcs_copy);
+    for (seen_pcs, seen_pcs_copy) |*src, *dst| dst.* = @as(*const volatile usize, @ptrCast(src)).*;
     const n_runs = @atomicLoad(usize, &cov_header.n_runs, .monotonic);
     const unique_runs = @atomicLoad(usize, &cov_header.unique_runs, .monotonic);
     {
@@ -304,7 +303,7 @@ pub fn sendUpdate(
         };
         var iovecs: [2][]const u8 = .{
             @ptrCast(&header),
-            @ptrCast(seen_pcs),
+            @ptrCast(seen_pcs_copy),
         };
         try socket.writeMessageVec(&iovecs, .binary);
