diff --git a/lib/std/http/Server.zig b/lib/std/http/Server.zig
index 7820ad6d0d..7079525786 100644
--- a/lib/std/http/Server.zig
+++ b/lib/std/http/Server.zig
@@ -628,7 +628,10 @@ pub const Request = struct {
         if (keep_alive and request.head.keep_alive) switch (r.state) {
             .received_head => {
                 if (request.head.method.requestHasBody()) {
-                    assert(request.head.transfer_encoding != .none or request.head.content_length != null);
+                    if (request.head.transfer_encoding == .none and request.head.content_length == null) {
+                        r.state = .closing;
+                        return false;
+                    }
                     const reader_interface = request.readerExpectContinue(&.{}) catch return false;
                     _ = reader_interface.discardRemaining() catch return false;
                     assert(r.state == .ready);
