diff --git a/lib/std/zig/Server.zig b/lib/std/zig/Server.zig
index 96679d4fc3..d120b62dfc 100644
--- a/lib/std/zig/Server.zig
+++ b/lib/std/zig/Server.zig
@@ -301,6 +301,10 @@ pub const TestMetadata = struct {
 };
 
 pub fn serveTestMetadata(s: *Server, test_metadata: TestMetadata) !void {
+    if (test_metadata.expected_panic_msgs.len != test_metadata.names.len) {
+        return error.InvalidArgument;
+    }
+
     const header: OutMessage.TestMetadata = .{
         .tests_len = @intCast(test_metadata.names.len),
         .string_bytes_len = @intCast(test_metadata.string_bytes.len),
