diff --git a/lib/c/strings.zig b/lib/c/strings.zig
index 111da9ad0f..ac564e132b 100644
--- a/lib/c/strings.zig
+++ b/lib/c/strings.zig
@@ -45,7 +45,8 @@ fn rindex(str: [*:0]const c_char, value: c_int) callconv(.c) ?[*:0]c_char {
 }
 
 fn firstBitSet(comptime T: type, value: T) T {
-    return @bitSizeOf(T) - @clz(value);
+    if (value == 0) return 0;
+    return @as(T, @ctz(value)) + 1;
 }
 
 fn ffs(i: c_int) callconv(.c) c_int {
